core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala

```diff
@@ -1995,6 +2000,61 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with B
     assert(!normalTSM.runningTasksSet.contains(taskId))
   }
 
+  test("SPARK-36575: Executor lost cause task hang") {
+    val taskScheduler = setupScheduler()
+
+    val resultGetter = new TaskResultGetter(sc.env, taskScheduler) {
+      override protected val getTaskResultExecutor: ExecutorService =
+        ThreadUtils.newDaemonFixedThreadPool(1, "task-result-getter")
+      def taskResultExecutor() : ExecutorService = getTaskResultExecutor
+    }
+    taskScheduler.taskResultGetter = resultGetter
+
+    val workerOffers = IndexedSeq(new WorkerOffer("executor0", "host0", 1),
+      new WorkerOffer("executor1", "host1", 1))
+    val task1 = new ShuffleMapTask(1, 0, null, new Partition {
+      override def index: Int = 0
+    }, Seq(TaskLocation("host0", "executor0")), new Properties, null)
+
+    val task2 = new ShuffleMapTask(1, 0, null, new Partition {
+      override def index: Int = 0
+    }, Seq(TaskLocation("host1", "executor1")), new Properties, null)
+
+    val taskSet = new TaskSet(Array(task1, task2), 0, 0, 0, null, 0)
+
+    taskScheduler.submitTasks(taskSet)
+    val taskDescriptions = taskScheduler.resourceOffers(workerOffers).flatten
+    assert(2 === taskDescriptions.length)
+
+    val ser = sc.env.serializer.newInstance()
+    val directResult = new DirectTaskResult[Int](ser.serialize(1), Seq(), Array.empty)
+    val resultBytes = ser.serialize(directResult)
+
+    // make getTaskResultExecutor busy
+    import scala.language.reflectiveCalls
+    resultGetter.taskResultExecutor().submit( new Runnable {
+      override def run(): Unit = Thread.sleep(100)
+    })
+
+    // task1 finished
+    taskScheduler.statusUpdate(
+      tid = taskDescriptions(0).taskId,
+      state = TaskState.FINISHED,
+      serializedData = resultBytes
+    )
+
+    // mark executor heartbeat timed out
+    taskScheduler.executorLost(taskDescriptions(0).executorId, ExecutorProcessLost("Executor " +
+      "heartbeat timed out"))
+
+    // Wait a while until all events are processed
+    Thread.sleep(100)
```

Comment:
Using thread.sleep's here will result in flakey tests.
Instead, mock the relevant methods and use latches to control behavior.

Reply:
I had optimized it, you can review it if you have time.

Reply:
I had optimized it, you can review it if you have time.

> I don't understand how the issue here leads to the stage hang. From what I see now, the task is considered a failed task first (due to executor loss) and reenqueue to the pending list, then it's considered a successful task later (due to statusupdate event). This does cause some inconsistent status but doesn't seem to cause the stage hang. IIRC, the stage can also complete even if the pending list isn't empty.
> 
> I probably missing something here. Could you elaborate more on how it causes stage hang?

The issue will  cause `TaskSetManager.successful` wrong result, mark the partition index true.  And `TaskSetManager.dequeueTaskFromList` only dequeue the unsuccessful index (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L313), so this partition index will be never scheduled again.
I also had update unit test., add assert code `assert(false == taskSetManager.successful(taskDescriptions(0).index))`.

Reply:
> The issue will cause TaskSetManager.successful wrong result, mark the partition index true. And TaskSetManager.dequeueTaskFromList only dequeue the unsuccessful index (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L313), so this partition index will be never scheduled again.


As I mentioned, the stage can also complete even if the pending list isn't empty. I think it can still complete as long as `tasksSuccessful == numTasks`. I still can't get why it can lead to stage hang.

Reply:
Based on your question, I realize that some details are indeed missing. So reorganize the whole process based on the log. Finally, it was found that `TaskSetManager.executorLost` was executed twice, and the second time resulted in `tasksSuccessful -= 1`, resulting in `tasksSuccessful` always less than `numTasks` .
The reason `TaskSetManager.executorLost` is executed twice is because `Heartbeat.expireDeadHosts` executes `TaskSchedulerImpl.executorLost` for the first time, and then executes `sc.killAndReplaceExecutor`. This issue occurred in version 2.3.3. `Heartbeat.expireDeadHosts` had adjusted the logic of the master branch, so I am not sure whether the stage hang issue was existed in new version. BTW,  I updated the JIRA description, added more detail log.

Reply:
Ok, I checked the logs in JIRA with the 2.3 code. Now I can guess how `TaskSetManager.executorLost` can be called twice. The first time is called by `TaskSchedulerImpl.executorLost` in `HeartbeatReceiver`, and  the second time could be triggered by `onDisconnect -> removeExecutor -> executorLost` due to `sc.killAndReplaceExecutor`. And the second time can only take effect when a `resourceOffer` request happens before it so that the `executorId` can be added back to `executorIdToHost`.  But, as you can see,

https://github.com/apache/spark/blob/75cc3b2da9ee0b51ecf0f13169f2b634e36a60c4/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L936-L949

`tasksSuccessful -= 1` only happens when `successful(index)=true`. And after first `executorLost`, `successful(index)` is already `false`. So, I think `tasksSuccessful -= 1` wouldn't happen the second time. 

I hope you can provide the complete or more logs if possible so that we can figure out the what's the real issue here. For now, I still can't see how the stage gets hang.

Beside, in Master, `TaskSchedulerImpl.executorLost` is no longer directly called in `HeartbeatReceiver`. Instead, it's delegated to the `CoarseGrainedSchedulerBackend`. And since `sc.killAndReplaceExecutor` would also be delegated to the `CoarseGrainedSchedulerBackend` at the end, thus, executor data would be synchronized by `CoarseGrainedSchedulerBackend`. So, I think call `executorLost` twice won't happen in Master.

Reply:
> `tasksSuccessful -= 1` only happens when `successful(index)=true`. And after first `executorLost`, `successful(index)` is already `false`. So, I think `tasksSuccessful -= 1` wouldn't happen the second time.
> 
There is a finished event of that task between the first `executorLost` and the second `executorLost`, so `successful(index)` was `true` when second `executorLost` was called.

Reply:
> Finally, it was found that TaskSetManager.executorLost was executed twice, and the second time resulted in tasksSuccessful -= 1, resulting in tasksSuccessful always less than numTask.

Then, how could it result in "tasksSuccessful always less than numTask" if there's a task finish event? The event, IIUC, is a successful event, which will do `tasksSuccessful +=1` first, right?

Reply:
> 
>  The event, IIUC, is a successful event, which will do `tasksSuccessful +=1` first, right?

Yes. But the second `executorLost` made `tasksSuccessful -= 1`. With `successful(index)` was true now,  the task would never been scheduled again,  which resulted in "tasksSuccessful always less than numTask"

Reply:
Wouldn't the second `executorLost` reset `successful(index)` to false?

Reply:
You are right, `successful(index)` is indeed no problem, it is my mistake. But I also saw `copiesRunning(index) -= 1` were processed in `executorLost` , twice `executorLost`  made `copiesRunning(index)` to `-1`.  So this may be the real problem, am i right ?

Reply:
Ok, that's true. I think the root cause is due to `copiesRunning(index) = -1`. But the problem exists in 2.3. In Master, as I mentioned above, twice `executorLost` can no longer happen.

Reply:
Yes. But it cause` TaskSetManager.successful` and `TaskSetManager.tasksSuccessful` wrong result in Matser, I think it's better to fix it.

Reply:
Yea, but could you rephrase the PR description first?